### PR TITLE
Module signatures

### LINKED ISSUE #1658: Module signatures supporting public/private syntax and imports
This would enable more incremental parsing.

We want to be able to:

-   Declare that a piece of syntax is `private`, so it cannot be used in either (i) modules which import the module it's declared in, or (ii) programs.
-   Declare that you are importing a module `A` into module `B` in a `private` manner, meaning that the module `B` may use the syntax of module `A`, but any module importing `B` may not use the syntax of module `A`.

This will cause problems with the existing lexer generation, because currently we only generate a single lexer for the entire definition. Roughly we can start with "the lexer should ignore the public/private status, while the parser should respect them".

For reference, we should profile the processes of (i) lexer generation and (ii) grammar generation. Currently we think that (i) takes on the order of seconds, and (ii) takes on the order of milliseconds (for "normal" definitions).

### PR DESCRIPTION
**Pull request by dwightguth**

### CODE REVIEW

**Comment by nishantjr** [comment]
@dwightguth Could we add a test for this case:

```
module C1 [private]
  imports B
endmodule
```

I think this should be equivalent to:

```
module C1 [private]
  imports [private] B
endmodule
```

and we also need syntax for:

```
module C1 [private]
  imports public B
endmodule
```

**Comment by dwightguth** [comment]
Currently my implementation treats `imports` as always public and `imports private` as always private. I guess we could make it so that we had both `imports private` and `imports public` and that `imports` was treated as having the same visibility as the module itself. I tend to agree that's more intuitive. On the other hand, imports, unlike other syntax, are always at the top of the module, so you would be unlikely to forget and give an import the wrong accessibility, which is the reason why we introduced the module attribute in the first place. I'm open to the change, but I'd like to hear what @radumereuta and @ehildenb think the intuitive behavior here is as well.

File: k-distribution/tests/regression-new/checks/signature.k.out [hunk 1]
```diff
@@ -0,0 +1,28 @@
+[Error] Inner Parser: Parse error: unexpected token 'foo'.
+	Source(signature.k)
+	Location(24,8,24,11)
+[Error] Inner Parser: Parse error: unexpected token 'fu'.
+	Source(signature.k)
+	Location(26,8,26,10)
+[Error] Inner Parser: Parse error: unexpected token 'baz'.
+	Source(signature.k)
+	Location(28,8,28,11)
```
Comment (ehildenb) [inline k-distribution/tests/regression-new/checks/signature.k.out:9]: Can we make these error messages more informative easily? Something like also saying "Symbol 'baz' is part of the private signature of module 'A', and so cannot be used directly in module 'B'"?

Comment (dwightguth) [reply k-distribution/tests/regression-new/checks/signature.k.out:9]: Not easily. We would have to have some way of recognizing that it parsed as a "baz" and then rejecting that parse, which is going to be a lot less efficient than simply removing it from the grammar entirely.

File: kernel/src/main/java/org/kframework/parser/inner/generator/RuleGrammarGenerator.java [hunk 2]
```diff
@@ -129,9 +132,10 @@ public Module getRuleGrammar(Module mod) {
...
         // import CONFIG-CELLS in order to parse cells specific to configurations
         Module newM = new Module( mod.name() + "-" + CONFIG_CELLS
-                                , Set(mod, baseK.getModule(K).get(), baseK.getModule(CONFIG_CELLS).get(), baseK.getModule(DEFAULT_LAYOUT).get())
-                                , Set()
-                                , Att()
+                                , (scala.collection.Set<Module>) mod.publicImports().$bar(Set(baseK.getModule(K).get(), baseK.getModule(CONFIG_CELLS).get(), baseK.getModule(DEFAULT_LAYOUT).get()))
+                                , mod.privateImports()
+                                , mod.localSentences()
+                                , mod.att()
                                 );
```
Comment (ehildenb) [inline kernel/src/main/java/org/kframework/parser/inner/generator/RuleGrammarGenerator.java]: This pattern than `getConfigGrammar` uses seems like it could be extracted into a single function `getGrammar`, which takes as input `Module mod, String cellName`, then `getConfigGrammar(Module mod) => getGramma(mod, mod.name() + "-" + CONFIG_CELLS)`, for instance. That would deduplicate `getRuleGrammar` and `getConfigGrammar`

File: kernel/src/main/java/org/kframework/parser/inner/generator/RuleGrammarGenerator.java [hunk 3]
```diff
@@ -179,15 +187,19 @@ public static boolean isParserSort(Sort s) {
...
     }
 
     public static ParseInModule getCombinedGrammar(Module mod, boolean strict, boolean timing, FileUtil files) {
-      return getCombinedGrammar(mod, strict, timing, false, files);
+      return getCombinedGrammar(mod, strict, timing, false, false, files);
+    }
+
+    public static ParseInModule getCombinedGrammar(Module mod, boolean strict, boolean timing, FileUtil files, boolean forGlobalScanner) {
+      return getCombinedGrammar(mod, strict, timing, false, forGlobalScanner, files);
     }
```
Comment (ehildenb) [inline kernel/src/main/java/org/kframework/parser/inner/generator/RuleGrammarGenerator.java]: Can we check which of these overloads are used directly and remove the ones that aren't (if there are any)? Maybe we don't need the one without `boolean forGlobalScanner` anymore?

File: kernel/src/main/java/org/kframework/parser/inner/kernel/Grammar.java [hunk 4]
```diff
@@ -58,6 +58,14 @@
     /** The set of "root" NonTerminals */
     private BiMap<String, NonTerminal> startNonTerminals = HashBiMap.create();
 
+    public NonTerminal nullNT() {
+      return get("_Null");
+    }
+
+    public Grammar() {
+      add(new NonTerminal("_Null"));
+    }
```
Comment (ehildenb) [inline kernel/src/main/java/org/kframework/parser/inner/kernel/Grammar.java:67]: We're sure these "special" nonterminals cannot show up anywhere in a user's grammar?

Comment (dwightguth) [reply kernel/src/main/java/org/kframework/parser/inner/kernel/Grammar.java:67]: I'm sure. Sort names can't begin with an underscore.

File: kore/src/main/scala/org/kframework/definition/outer-ext.scala [hunk 5]
```diff
@@ -56,12 +56,14 @@ object FlatModule {
...
+        val f = (i: Import) => memoization.getOrElse(i.name
           // if can't find the Module in memoization, build a new one
           , toModuleRec(allModules.find(f => f.name.equals(i.name))
               .getOrElse(throw KEMException.compilerError("Could not find module: " + i.name, i))
-            , visitedModules :+ m)))
-        val newM = new Module(m.name, newImports, m.localSentences, m.att)
+            , visitedModules :+ m))
+        val newPublicImports = m.imports.filter(_.isPublic).map(f)
+        val newPrivateImports = m.imports.filter(!_.isPublic).map(f)
+        val newM = new Module(m.name, newPublicImports, newPrivateImports, m.localSentences, m.att)
```
Comment (ehildenb) [inline kore/src/main/scala/org/kframework/definition/outer-ext.scala:66]: This construction pattern (split into public/private sentences, then build new Module type) seems common. Can we extract a "smart constructor" for `Module` which encapsulates that logic (so you just pass in the full set of sentences, and it does the split and constructs the module with the split sentences)?

File: k-distribution/tests/regression-new/checks/signature.k.out (hunk 1 above)
Comment (radumereuta) [reply k-distribution/tests/regression-new/checks/signature.k.out:9]: I agree with Dwight. One benefit of hiding syntax is that the parser doesn't have to handle those cases, which might lead to fewer ambiguities and faster parsing.
It's doable, but it adds additional complexity and makes parsing equally slow.
Then raises the question, if you are doing it for imported private syntax, why not do it for parallel modules...

**Comment by nishantjr** [comment]
@dwightguth Could you also add documentation to the manual please.

File: kernel/src/main/java/org/kframework/parser/inner/generator/RuleGrammarGenerator.java (hunk 3 above)
Comment (radumereuta) [reply kernel/src/main/java/org/kframework/parser/inner/generator/RuleGrammarGenerator.java]: I'm confused about `forGlobalScanner` as well. Why is it needed?

**Review by radumereuta — CHANGES_REQUESTED** [review]
This is very large, and I will have to go through it again in more detail. But for now, I found a test that fails to pass:
```
module TEST
  syntax Foo ::= foo() // <- B may not use this symbol

  syntax Foo ::= bam() [public] // <- B may use this symbol

  syntax Foo ::= fu() [private] // <- B may not use this symbol

  syntax Foo ::= "bar" [public, token] // <- B may use this token
               | r"ba." [token, prec(3)]      // <- B may not use this token

  rule foo() => .K
  rule bam() => .K
  rule fu() => .K
  rule bar => .K
  rule baz => .K
endmodule
```
```
$ kompile test.k
[Error] Internal: Uncaught exception thrown of type MatchError.
Please rerun your program with the --debug flag to generate a stack trace, and
file a bug report at https://github.com/kframework/k/issues (MatchError:
(LiteralC(bar),Lblbam'LParRParUnds'TEST'Unds'Foo{}) (of class scala.Tuple2))

$ kompile test.k --backend haskell
Error:
  module 'TEST':
  symbol 'Lblbam'LParRParUnds'TEST'Unds'Foo' declaration (definition.kore 113:10):
    Cannot define constructor Lblbam'LParRParUnds'TEST'Unds'Foo for sort SortFoo{}: sort has domain values.
[Error] Critical: Haskell backend reported errors validating compiled
definition.
Examine output to see errors.
```

**Comment by radumereuta** [comment]
The issue fails on master also. Very strange.
If this ends up being outside the scope of this PR, please add a test that checks for the positive case, where you have public/private syntax, and compilation passes.

**Comment by radumereuta** [comment]
> Currently my implementation treats `imports` as always public and `imports private` as always private. I guess we could make it so that we had both `imports private` and `imports public` and that `imports` was treated as having the same visibility as the module itself. I tend to agree that's more intuitive. On the other hand, imports, unlike other syntax, are always at the top of the module, so you would be unlikely to forget and give an import the wrong accessibility, which is the reason why we introduced the module attribute in the first place. I'm open to the change, but I'd like to hear what @radumereuta and @ehildenb think the intuitive behavior here is as well.

Sounds like a good idea. It gives you more control over syntax visibility. If it's not a big change, I'm in favor.

File: kernel/src/main/java/org/kframework/parser/inner/generator/RuleGrammarGenerator.java (hunk 3 above)
Comment (dwightguth) [reply kernel/src/main/java/org/kframework/parser/inner/generator/RuleGrammarGenerator.java]: All of these overloads are used somewhere. In response to Radu's question, it's needed so that the scanner we generate to parse rules contains /all/ of the syntax and not just the syntax visible to the main module.

File: kore/src/main/scala/org/kframework/definition/outer-ext.scala (hunk 5 above)
Comment (dwightguth) [reply kore/src/main/scala/org/kframework/definition/outer-ext.scala:66]: In point of fact, while there are three places in this PR that do something similar like this, they are enough different that they would not benefit from refactoring out a common function.

**Comment by dwightguth** [comment]
@radumereuta, please create a separate issue for the MatchError exception you encountered, as it is outside the scope of this PR.

**Comment by dwightguth** [comment]
I have not yet written the manual segment, but the code itself has had all of the issues everyone raised addressed in some form or other. I still will want someone to review the documentation once I write it, but the code ought to be ready for re-review now.

**Comment by nishantjr** [comment]
Looks good. Thanks!

File: kernel/src/main/java/org/kframework/parser/inner/generator/RuleGrammarGenerator.java (hunk 3 above)
Comment (radumereuta) [reply kernel/src/main/java/org/kframework/parser/inner/generator/RuleGrammarGenerator.java]: If I understand correctly, you are trying to optimize the case when you are parsing `STDOUT-STREAM` and `STDIN-STREAM` by reducing the number of productions included in the scanner. IMO the possible gains in speed don't merit adding a new parameter here. So I would like either:
- remove the option. The benefits don't outweigh the added code complexity.
- add inline documentation explaining the functionality. I spent more than half an hour trying to understand why this is needed and how it works.

If I'm getting this wrong, please correct me.

**Review by radumereuta — CHANGES_REQUESTED** [review]
2 changes
- `forGlobalScanner`  (see my other comment)
- please add a test with the new public/private syntax that kompiles w/o errors (a check with the empty output would be sufficient)

Question: you were talking about changes to domains.md to include public/private syntax. I don't see any changes to that file. Am I remembering wrong?

Other than that looks pretty good.

File: kernel/src/main/java/org/kframework/parser/inner/generator/RuleGrammarGenerator.java (hunk 3 above)
Comment (dwightguth) [reply kernel/src/main/java/org/kframework/parser/inner/generator/RuleGrammarGenerator.java]: No see, if the main scanner for rules doesn't include private syntax, then we will need to generate a new scanner for every module containing any private syntax, which will be very expensive.

**Comment by nishantjr** [comment]
> Question: you were talking about changes to domains.md to include public/private syntax. I don't see any changes to that file. Am I remembering wrong?

@radumereuta That would be a non-backward compatible change. We should perhaps do that when we bump major versions?

**Comment by dwightguth** [comment]
@radumereuta the changes to domains.md are in a separate pr in order to simplify things, because as Nishant says, they are not backwards compatible. I have commented on your comments about the scanner, but I can't make that change without causing serious problems. I will add the test you ask for though.

File: kernel/src/main/java/org/kframework/parser/inner/generator/RuleGrammarGenerator.java (hunk 3 above)
Comment (radumereuta) [reply kernel/src/main/java/org/kframework/parser/inner/generator/RuleGrammarGenerator.java]: I understand this. Why isn't this the default for every other scanner? Why do you need to make it an option?

Comment (dwightguth) [reply kernel/src/main/java/org/kframework/parser/inner/generator/RuleGrammarGenerator.java]: This is the only place where scanners are created that are used non-modularly for other modules. Every other place where a ParseInModule is being constructed, we either don't use the scanner that comes with that object, or else we use that scanner only to scan that module, in which case we don't /want/ to include the private syntax in the scanner.

Comment (radumereuta) [reply kernel/src/main/java/org/kframework/parser/inner/generator/RuleGrammarGenerator.java]: That sounds reasonable. Please add that explanation as a code comment. It's going to be useful in the future.

File: USER_MANUAL.md [hunk 6]
```diff
@@ -537,6 +588,24 @@ automatically. It will also automatically sort associative-commutative
...
+module in which it is declared; it is not visible from modules that import
+that module.
+
+You can also change the default visibility of a module with the `private`
+attribute, when it is placed directly on a module. A module with the `private`
+attribute has all syntax `private` by default; this can be overridden on
+specific sentences with the `public` attribute.
+
+Note that the `private` module attribute also changes the default visiblity
+of imports; please refer to the appropriate section elsewhere in the manual
```
Comment (radumereuta) [inline USER_MANUAL.md:606]: Don't you have a link or reference instead of just _elsewhere_?
Sounds dismissive.

**Review by radumereuta — APPROVED** [review]
Just one comment. Otherwise looks pretty good. Let's start testing it.

File: USER_MANUAL.md (hunk 6 above)
Comment (dwightguth) [reply USER_MANUAL.md:606]: I'm open to including a link if you know how...

File: USER_MANUAL.md [hunk 7]
```diff
@@ -537,6 +588,24 @@ automatically. It will also automatically sort associative-commutative
...
+for more details.
+
```
Comment (nishantjr) [inline USER_MANUAL.md:630]: I'm not sure how to suggest code blocks, but I think we should add an example.

```suggestion
We give an example of possible usage below:

module WIDGET
  syntax Widget ::= gadget()
endmodule

module FOO [private]
  imports WIDGET                         // <- B may not use `gadget()`
  syntax Foo ::= foo()                   // <- B may not use this symbol
  syntax Foo ::= bam() [public]          // <- B may use this symbol
  syntax Foo ::= fu() [private]          // <- B may not use this symbol
  syntax Foo ::= "bar" [public, token]   // <- B may use this token
               | r"ba." [token, prec(3)] // <- B may not use this token
endmodule

module B
  imports private FOO                    // <- B may use the _public_ syntax of A, but C may not
  syntax Bar ::= a() [public]            // <- C may use this symbol
               | b() [private]           // <- C may not use this symbol
endmodule

module C
  imports B
  imports WIDGET                         // C may now use WIDGET's public interface
endmodule

```

Comment (ehildenb) [reply USER_MANUAL.md:630]: I agree that we should add an example like this.
